src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,192 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
+  // Convert a + a + ... + a into a*n
+  Node* serial_additions = convert_serial_additions(phase, can_reshape, bt);
+  if (serial_additions != nullptr) {
+    return serial_additions;
+  }
+
   return AddNode::Ideal(phase, can_reshape);
 }
 
+// Convert a + a + ... + a into a*n
+Node* AddNode::convert_serial_additions(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  if (is_optimized_multiplication()) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  int in1_op = in1->Opcode();
+
+  bool matched = false;
+  jlong multiplier;
+
+  // Convert (a + a) + a to 3 * a
+  // Look for LHS pattern: AddNode(a, a)
+  if (in1_op == Op_Add(bt) && in1->in(1) == in1->in(2)) {
+    if (in1->in(1) != in2) {
+      return nullptr;
+    }
+
+    multiplier = 2;
+    matched = true;
+  }
+
+  // Convert (a << CON) + a to a * ((1 << CON) + 1))
+  // Look for LHS pattern: LShiftNode(a, CON)
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (in1_op == Op_LShift(bt) && in1->in(2)->is_Con()) {
+    Node* base = in1->in(1);
+    if (base != in2) {
+      return nullptr;
+    }
+
+    Node* con = in1->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
+      return nullptr;
+    }
+
+    multiplier = (jlong) 1 << con->get_integer_as_long(con_bt);
+    matched = true;
+  }
+
+  // Convert (CON * a) + a to (CON + 1) * a
+  // Look for LHS patterns:
+  //     - MulNode(CON, a)
+  //     - MulNode(a, CON)
+  // This optimization technically only produces MulNode(CON, a), but we might as well transform (a * CON) + a, too.
+  if (in1_op == Op_Mul(bt) && (in1->in(1)->is_Con() || in1->in(2)->is_Con())) {
+    Node* base = in1->in(1)->is_Con() ? in1->in(2) : in1->in(1);
+    if (base != in2) {
+      return nullptr;
+    }
+
+    Node* con = in1->in(1)->is_Con() ? in1->in(1) : in1->in(2);
+    BasicType con_bt = phase->type(con)->basic_type();
+    if (con_bt == T_VOID) { // const could potentially be void type
+      return nullptr;
+    }
+
+    multiplier = con->get_integer_as_long(con_bt);
+    matched = true;
+  }
+
+  // Due to potential power-of-2 multiplication optimization, e.g., ((a << CON1) + (a << CON2)) + a,
+  // we also look for LHS patterns:
+  //     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+  if (in1_op == Op_Add(bt) && in1->isa_Add()->is_optimized_multiplication()) {
+    Node* lhs = in1->in(1);
+    Node* rhs = in1->in(2);
+
+    Node* lhs_base = lhs->is_LShift() ? lhs->in(1) : lhs;
+    Node* rhs_base = rhs->is_LShift() ? rhs->in(1) : rhs;
+
+    Node* nodes[] = {lhs, rhs};
+    Node* bases[] = {lhs_base, rhs_base};
+    jlong multipliers[] = {1, 1};
+
+    // AddNode(LShiftNode(a, CON1), LShiftNode(a, CON2)/a)
+    // AddNode(LShiftNode(a, CON1)/a, LShiftNode(a, CON2))
+    for (int i = 0; i < 2; i++) {
```

Comment:
I wouldn't use a loop here. I would put the loop body into its own method and call it twice, once with `lhs`, `lhs_base` as arguments, once with `rhs`, `rhs_base`.

Reply:
I refactored even further to combine checking for optimized `mul`s and extracting multipliers to use the same logic. This code is now obsolete.